pkg/hubble/monitor/consumer.go

```diff
@@ -27,70 +28,109 @@ type Observer interface {
+			strings.ToLower(flowpb.LostEventSource_OBSERVER_EVENTS_QUEUE.String())),
 	}
 	return mc
 }
 
 // sendEventQueueLostEvents tries to send the current value of the lost events
 // counter to the observer. If it succeeds to enqueue a notification, it
-// resets the counter.
-func (c *consumer) sendNumLostEvents() {
+// resets the counter. Returns a boolean indicating whether the notification
+// has been successfully sent.
+func (c *consumer) sendNumLostEvents() bool {
 	c.lostLock.Lock()
 	defer c.lostLock.Unlock()
 	// check again, in case multiple
 	// routines contended the lock
 	if c.numEventsLost == 0 {
-		return
+		return true
 	}
 
-	numEventsLostNotification := &observerTypes.MonitorEvent{
-		UUID:      uuid.New(),
-		Timestamp: time.Now(),
-		NodeName:  nodeTypes.GetAbsoluteNodeName(),
-		Payload: &observerTypes.LostEvent{
-			Source:        observerTypes.LostEventSourceEventsQueue,
-			NumLostEvents: c.numEventsLost,
-		},
+	if c.cachedLostNotification == nil {
+		c.cachedLostNotification = c.newEvent(func() interface{} {
+			return &observerTypes.LostEvent{
+				Source:        observerTypes.LostEventSourceEventsQueue,
+				NumLostEvents: c.numEventsLost,
+			}
+		})
+	} else {
+		c.cachedLostNotification.Timestamp = time.Now()
+		c.cachedLostNotification.Payload.(*observerTypes.LostEvent).NumLostEvents = c.numEventsLost
 	}
+
 	select {
-	case c.observer.GetEventsChannel() <- numEventsLostNotification:
+	case c.observer.GetEventsChannel() <- c.cachedLostNotification:
 		// We now now safely reset the counter, as at this point have
 		// successfully notified the observer about the amount of events
-		// that were lost since the previous LostEvent message
+		// that were lost since the previous LostEvent message. Similarly,
+		// we reset the cached notification, so that a new one is created
+		// the next time.
 		c.numEventsLost = 0
+		c.cachedLostNotification = nil
+		return true
 	default:
 		// We do not need to bump the numEventsLost counter here, as we will
 		// try to send a new LostEvent notification again during the next
 		// invocation of sendEvent
+		return false
 	}
 }
 
 // sendEvent enqueues an event in the observer. If this is not possible, it
 // keeps a counter of lost events, which it will regularly try to send to the
 // observer as well
-func (c *consumer) sendEvent(event *observerTypes.MonitorEvent) {
+func (c *consumer) sendEvent(payloader func() interface{}) {
 	if c.numEventsLost > 0 {
-		c.sendNumLostEvents()
+		if !c.sendNumLostEvents() {
+			// We just failed sending the lost notification, hence it doesn't
+			// make sense to try and send the actual event, as we'll most
+			// likely fail as well.
+			c.countDroppedEvent()
+			return
+		}
 	}
+
 	select {
-	case c.observer.GetEventsChannel() <- event:
+	case c.observer.GetEventsChannel() <- c.newEvent(payloader):
 	default:
 		c.countDroppedEvent()
 	}
 }
 
+func (c *consumer) newEvent(payloader func() interface{}) *observerTypes.MonitorEvent {
```

Comment:
nit: this `func` doesn't need to be a struct method right?

Reply:
Hmm, it actually needs to access the `uuider` field of `consumer`, which is added in e30568eb560a91e851d41a9fd58a61f5055b063c. And I've created it as a struct method from the beginning to avoid having to change it in the subsequent commits.